examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+ *
+ * @param    currentmA
+ */
+Status EnergyEvseDelegate::HwSetCableAssemblyLimit(int64_t currentmA)
+{
+    if (currentmA < kMinimumChargeCurrent || currentmA > kMaximumChargeCurrent)
+    {
+        return Status::ConstraintError;
+    }
+
+    /* there is no attribute to store this so store in private variable */
+    mCableAssemblyCurrentLimit = currentmA;
+
+    return this->ComputeMaxChargeCurrentLimit();
+}
+
+/**
+ * @brief    Called by EVSE Hardware to indicate a fault
+ *
+ * @param    StateEnum
+ */
+Status EnergyEvseDelegate::HwSetState(StateEnum state)
+{
+    /* the only allowed states that the EVSE hardware can set are:
+     *  kNotPluggedIn
+     *  kPluggedInNoDemand
+     *  kPluggedInDemand
+     */
+    switch (state)
+    {
+    case StateEnum::kNotPluggedIn:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+    case StateEnum::kPluggedInNoDemand:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+    case StateEnum::kPluggedInDemand:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+
+    default:
+        /* All other states should be managed by the Delegate */
+        // TODO (assert?)
+        break;
+    }
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called by EVSE Hardware to indicate a fault
+ *
+ * @param    FaultStateEnum
+ */
+Status EnergyEvseDelegate::HwSetFault(FaultStateEnum fault)
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::Fault()");
+
+    if (fault == FaultStateEnum::kNoError)
+    {
+        /* Update State to previous state */
+        // TODO: need to work out the logic here!
+
+        /* Update SupplyState to previous state */
+    }
+    else
+    {
+        /* Update State & SupplyState */
+        SetState(StateEnum::kFault);
+        SetSupplyState(SupplyStateEnum::kDisabledError);
+    }
+
+    /* Update FaultState */
+    SetFaultState(fault);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called by EVSE Hardware to Send a RFID event
+ *
+ * @param    ByteSpan RFID tag value (max 10 octets)
+ */
+Status EnergyEvseDelegate::HwSetRFID(ByteSpan uid)
+{
+    Events::Rfid::Type event{ .uid = uid };
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(event, mEndpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "[Notify] Unable to send notify event: %s [endpointId=%d]", error.AsString(), mEndpointId);
+        return Status::Failure;
+    }
+
+    return Status::Success;
+}
+/**
+ * @brief    Called by EVSE Hardware to share the VehicleID
+ *
+ * This routine will make a copy of the string so the callee doesn't
+ * have to hold onto it forever.
+ *
+ * @param    CharSpan containing up to 32 chars.
+ */
+Status EnergyEvseDelegate::HwSetVehicleID(const CharSpan & newValue)
+{
+    if ((mVehicleID.IsNull()) || (strcmp(newValue.data(), mVehicleID.Value().data()) != 0))
```

Comment:
CharSpan is not null-terminated.  This right here is a buffer overrun.  You want something more like this:
```suggestion
    if ((mVehicleID.IsNull()) || !(newValue.data_equal(mVehicleID.Value()))
```
or however you want to place the parentheses here.  If I were writing this myself it would be:
```
    if (mVehicleID.IsNull() || !newValue.data_equal(mVehicleID.Value())
```

Reply:
Good spot - Fixed.
Also removed the ChipLogDetail %s which would print out extra chars if not null terminated.